### PR TITLE
[Batch Mode] Rdar 42314665 fix batch mode truncation bug

### PR DESCRIPTION
**Pull request by davidungar**
Batch mode must truncate the serialized diagnostics file of any primary which does not have errors specifically attached to it. However, some errors are not specific to any primary, and these must count as attached to every primary. This PR fixes the code to do the latter. It also refactors the FileSpecificDiagnosticConsumer to avoid future bugs.

It also includes tests for same, courtesy of @jrose-apple .

Resolves rdar://43033749

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
I bet we can find a better name for ‘ConsumerSpecificInformation’. ‘DiagosticFilter’? ...Range? ...Director?

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticConsumer.h [hunk 1]
```diff
@@ -136,32 +136,110 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
...
+    /*const*/ unsigned subconsumerIndex;
+
+  public:
     ConsumerSpecificInformation(const CharSourceRange range,
-                                DiagnosticConsumer *const consumer)
-        : range(range), consumer(consumer) {}
+                                unsigned subconsumerIndex)
+        : range(range), subconsumerIndex(subconsumerIndex) {}
+
+    Subconsumer &subconsumer(FileSpecificDiagnosticConsumer &c) const {
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticConsumer.h]: I see why you did this: it lets you keep the `subconsumerIndex` field private, and makes the code sort of look like there's a `subconsumer` computed property. But having to pass in the owner breaks the illusion. I think I'd prefer something like `operator[](const ConsumerSpecificInformation &)` on FileSpecificDiagnosticConsumer if you want to keep `subconsumerIndex` out of client code.

File: include/swift/AST/DiagnosticConsumer.h [hunk 2]
```diff
@@ -136,32 +136,110 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
...
+    /// The range of SourceLoc's for which diagnostics should be directed to
+    /// this subconsumer.
     /*const*/ CharSourceRange range;
-    /// The DiagnosticConsumer may be empty if those diagnostics are not to be
-    /// emitted.
-    DiagnosticConsumer * /*const*/ consumer;
-    bool hasAnErrorBeenEmitted = false;
 
+    /// Index into Subconsumers vector for this subconsumer.
+    /*const*/ unsigned subconsumerIndex;
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticConsumer.h]: Why can't you make these const?

File: lib/AST/DiagnosticConsumer.cpp [hunk 3]
```diff
@@ -171,22 +185,12 @@ void FileSpecificDiagnosticConsumer::handleDiagnostic(
...
   }
-  if (!consumerSpecificInfo.getValue()->consumer)
-    return; // Suppress non-primary diagnostic in batch mode.
-
-  consumerSpecificInfo.getValue()->consumer->handleDiagnostic(
-      SM, Loc, Kind, FormatString, FormatArgs, Info);
-  consumerSpecificInfo.getValue()->hasAnErrorBeenEmitted |=
-      Kind == DiagnosticKind::Error;
+  for (auto &subconsumer : Subconsumers)
+    subconsumer.handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs, Info);
```
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: This consolidation came out nicely!

File: lib/FrontendTool/FrontendTool.cpp [hunk 4]
```diff
@@ -1596,14 +1596,15 @@ createDispatchingDiagnosticConsumerIfNeeded(
...
+  SmallVector<FileSpecificDiagnosticConsumer::Subconsumer, 4> subconsumers;
 
   inputsAndOutputs.forEachInputProducingSupplementaryOutput(
       [&](const InputFile &input) -> bool {
-    if (auto subConsumer = maybeCreateSingleConsumer(input))
-      subConsumers.emplace_back(input.file(), std::move(subConsumer));
-    return false;
-  });
+        if (auto subconsumer = maybeCreateSingleConsumer(input))
+          subconsumers.emplace_back(FileSpecificDiagnosticConsumer::Subconsumer(
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: This is `emplace_back`, not `push_back`, precisely so that you don't have to name the type again.

File: lib/FrontendTool/FrontendTool.cpp [hunk 5]
```diff
@@ -1596,14 +1596,15 @@ createDispatchingDiagnosticConsumerIfNeeded(
...
-  SmallVector<FileSpecificDiagnosticConsumer::ConsumerPair, 4> subConsumers;
+  SmallVector<FileSpecificDiagnosticConsumer::Subconsumer, 4> subconsumers;
 
   inputsAndOutputs.forEachInputProducingSupplementaryOutput(
       [&](const InputFile &input) -> bool {
-    if (auto subConsumer = maybeCreateSingleConsumer(input))
-      subConsumers.emplace_back(input.file(), std::move(subConsumer));
-    return false;
-  });
+        if (auto subconsumer = maybeCreateSingleConsumer(input))
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: Probably want to rename this now, since "Subconsumer" refers to the whole info structure rather than just a subordinate DiagnosticConsumer.

File: unittests/AST/DiagnosticConsumerTests.cpp [hunk 6]
```diff
@@ -63,12 +63,15 @@ TEST(FileSpecificDiagnosticConsumer, SubConsumersFinishInOrder) {
   auto consumerUnaffiliated = llvm::make_unique<ExpectationDiagnosticConsumer>(
       consumerA.get(), None);
 
-  SmallVector<FileSpecificDiagnosticConsumer::ConsumerPair, 2> consumers;
-  consumers.emplace_back("A", std::move(consumerA));
-  consumers.emplace_back("", std::move(consumerUnaffiliated));
+  SmallVector<FileSpecificDiagnosticConsumer::Subconsumer, 2> consumers;
+  consumers.emplace_back(
+      FileSpecificDiagnosticConsumer::Subconsumer("A", std::move(consumerA)));
+  consumers.emplace_back(FileSpecificDiagnosticConsumer::Subconsumer(
```
Comment (jrose-apple) [inline unittests/AST/DiagnosticConsumerTests.cpp]: More `emplace_back`s that shouldn't need to change (all throughout the file).

**Review by jrose-apple — COMMENTED** [review]
The structure of this came out pretty good!

As far as ConsumerSpecificInformation goes, it looks like the name before it got generalized in #16526 was just "ConsumersOrderedByRangeEntry". That's not a wonderful name, but it does get the point across: it's just a mapping from source ranges to subconsumers (via an index). (I'm not suggesting using that verbatim, but just as a direction to go in.)

**Comment by davidungar** [comment]
Trying out `ConsumerWithCorrespondingRange` instead of `ConsumerSpecificInformation`.

File: lib/AST/DiagnosticConsumer.cpp [hunk 7]
```diff
@@ -83,9 +99,7 @@ void FileSpecificDiagnosticConsumer::computeConsumersOrderedByRange(
   std::sort(ConsumersOrderedByRange.begin(), ConsumersOrderedByRange.end(),
             [](const ConsumerSpecificInformation &left,
                const ConsumerSpecificInformation &right) -> bool {
-              auto compare = std::less<const char *>();
-              return compare(getRawLoc(left.range.getEnd()).getPointer(),
-                             getRawLoc(right.range.getEnd()).getPointer());
+              return left < right;
```
[the remark this answers is not in the document]
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: Thanks for being flexible. I do prefer it; the code seems to be telling me that these objects are ordered.

File: include/swift/AST/DiagnosticConsumer.h [hunk 8]
```diff
@@ -136,32 +136,110 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
...
+                std::unique_ptr<DiagnosticConsumer> consumer)
+        : inputFileName(inputFileName), consumer(std::move(consumer)) {}
+
+    void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
+                          DiagnosticKind Kind,
+                          StringRef FormatString,
+                          ArrayRef<DiagnosticArgument> FormatArgs,
+                          const DiagnosticInfo &Info) {
+      if (!getConsumer())
+        return; // Suppress non-primary diagnostic in batch mode.
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Dropped.

File: include/swift/AST/DiagnosticConsumer.h (hunk 1 above)
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: OK. Requires an accessor for `subconsumerIndex` to guard against mutation in public, but it does read better.

File: include/swift/AST/DiagnosticConsumer.h (hunk 2 above)
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Good catch! Back when there was a unique pointer in there, I don't think these things compiled, but I don't recall the details.

File: lib/AST/DiagnosticConsumer.cpp (hunk 3 above)
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: Yes, thank you for the original idea! Worked out to move some functions into the substructures.

File: lib/FrontendTool/FrontendTool.cpp (hunk 4 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Like magic!

File: lib/FrontendTool/FrontendTool.cpp (hunk 5 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Good! I'm trying `maybeCreateConsumerForDiagnosticsFrom`. Do you have something better?

File: unittests/AST/DiagnosticConsumerTests.cpp (hunk 6 above)
Comment (davidungar) [reply unittests/AST/DiagnosticConsumerTests.cpp]: Thanks. Fixed.

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: lib/FrontendTool/FrontendTool.cpp (hunk 5 above)
Comment (jrose-apple) [reply lib/FrontendTool/FrontendTool.cpp]: Ah, I was referring to the name of the result rather than the callback.

File: include/swift/AST/DiagnosticConsumer.h [hunk 9]
```diff
@@ -209,16 +286,15 @@ class FileSpecificDiagnosticConsumer : public DiagnosticConsumer {
...
 
   void computeConsumersOrderedByRange(SourceManager &SM);
 
   /// Returns nullptr if diagnostic is to be suppressed,
   /// None if diagnostic is to be distributed to every consumer,
   /// a particular consumer if diagnostic goes there.
-  Optional<ConsumerSpecificInformation *>
-  consumerSpecificInformationForLocation(SourceManager &SM,
-                                         SourceLoc loc) const;
+  Optional<ConsumerAndRange *> consumerAndRangeForLocation(SourceManager &SM,
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticConsumer.h]: I think at this point you could go all the way to `Optional<Subconsumer *>`, since there's no more interesting information in ConsumerAndRange.

File: lib/FrontendTool/FrontendTool.cpp [hunk 10]
```diff
@@ -1596,14 +1596,14 @@ createDispatchingDiagnosticConsumerIfNeeded(
...
-  SmallVector<FileSpecificDiagnosticConsumer::ConsumerPair, 4> subConsumers;
+  SmallVector<FileSpecificDiagnosticConsumer::Subconsumer, 4> subconsumers;
 
   inputsAndOutputs.forEachInputProducingSupplementaryOutput(
       [&](const InputFile &input) -> bool {
-    if (auto subConsumer = maybeCreateSingleConsumer(input))
-      subConsumers.emplace_back(input.file(), std::move(subConsumer));
-    return false;
-  });
+        if (auto subconsumer = maybeCreateConsumerForDiagnosticsFrom(input))
```
Comment (jrose-apple) [inline lib/FrontendTool/FrontendTool.cpp]: Comment still relevant here, so copying it in again: "subconsumer" is not the best name for the local variable, now that that refers to something else. (I was fine with the callback name; the type provides the rest of the information.)

File: include/swift/AST/DiagnosticConsumer.h (hunk 9 above)
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: It's a good idea, but runs into `const` trouble. The `[]` operator on `FileSpecificDiagnosticConsumer` is not `const` because we may be setting the emitted error bit in the subconsumer structure. `consumerAndRangeForLocation` is `const`. So I can see these choices:

1. `const_cast` `this` at the end of `consumerAndRangeForLocation`,
2. unconstify `consumerAndRangeForLocation`
3. let the `[]` operator be `const` and `const_cast` it to set the error bit.
4. status quo

What do you like? Or is there another choice?

File: lib/FrontendTool/FrontendTool.cpp (hunk 10 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Thanks. Changed to `consumer`.

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticConsumer.h (hunk 9 above)
Comment (jrose-apple) [reply include/swift/AST/DiagnosticConsumer.h]: Ah, hm. I guess (2) is my preferred choice, then, because while `consumerAndRangeForLocation` isn't doing any mutation itself it's making it possible for callers to mutate. I see how (4) avoids that problem, but it does so at the cost of making the ConsumerAndRange structure more prominent and making things less convenient for callers. But if you have another preference that's okay; I see the difficulty now.

File: lib/FrontendTool/FrontendTool.cpp (hunk 5 above)
Comment (davidungar) [reply lib/FrontendTool/FrontendTool.cpp]: Yes! Changed to `consumer`.

File: include/swift/AST/DiagnosticConsumer.h (hunk 9 above)
Comment (davidungar) [reply include/swift/AST/DiagnosticConsumer.h]: Implemented (2)--looks good!

**Comment by davidungar** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticConsumer.h [hunk 11]
```diff
@@ -209,16 +286,15 @@ class FileSpecificDiagnosticConsumer : public DiagnosticConsumer {
...
 
   void computeConsumersOrderedByRange(SourceManager &SM);
 
   /// Returns nullptr if diagnostic is to be suppressed,
   /// None if diagnostic is to be distributed to every consumer,
   /// a particular consumer if diagnostic goes there.
-  Optional<ConsumerSpecificInformation *>
-  consumerSpecificInformationForLocation(SourceManager &SM,
-                                         SourceLoc loc) const;
+  Optional<FileSpecificDiagnosticConsumer::Subconsumer *>
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticConsumer.h]: Nitpick: no need to qualify with FileSpecificDiagnosticConsumer within the implementation.

File: lib/AST/DiagnosticConsumer.cpp [hunk 12]
```diff
@@ -134,19 +146,17 @@ FileSpecificDiagnosticConsumer::consumerSpecificInformationForLocation(
...
+        return entry.endsAfter(loc);
+      });
 
   if (possiblyContainingRangeIter != ConsumersOrderedByRange.end() &&
-      possiblyContainingRangeIter->range.contains(loc)) {
-    return const_cast<ConsumerSpecificInformation *>(
-        possiblyContainingRangeIter);
+      possiblyContainingRangeIter->contains(loc)) {
+    auto *consumerAndRangeForLocation =
+        const_cast<ConsumerAndRange *>(possiblyContainingRangeIter);
```
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: Ah, this doesn't need to be `const_cast`-ed anymore.

File: lib/AST/DiagnosticConsumer.cpp [hunk 13]
```diff
@@ -196,19 +197,18 @@ bool FileSpecificDiagnosticConsumer::finishProcessing() {
...
 void FileSpecificDiagnosticConsumer::
-    tellSubconsumersToInformDriverOfIncompleteBatchModeCompilation() const {
+    tellSubconsumersToInformDriverOfIncompleteBatchModeCompilation() {
   if (!HasAnErrorBeenConsumed)
     return;
-  for (auto &info : ConsumersOrderedByRange) {
-    if (!info.hasAnErrorBeenEmitted && info.consumer)
-      info.consumer->informDriverOfIncompleteBatchModeCompilation();
-  }
+  for (auto &info : ConsumersOrderedByRange)
```
Comment (jrose-apple) [inline lib/AST/DiagnosticConsumer.cpp]: Maybe this should go in a separate commit, but this should also be iterating over Subconsumers instead of ConsumersOrderedByRange, so we get a stable iteration order between runs.

**Review by jrose-apple — APPROVED** [review]

File: lib/AST/DiagnosticConsumer.cpp (hunk 13 above)
Comment (davidungar) [reply lib/AST/DiagnosticConsumer.cpp]: Yes, I like it! Done!

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
Thanks for your review. This change came out nicely.
